drivers/xen/events.c

```diff
@@ -18,8 +19,15 @@ LOG_MODULE_REGISTER(xen_events);
 extern shared_info_t *HYPERVISOR_shared_info;
 
 static evtchn_handle_t event_channels[EVTCHN_2L_NR_CHANNELS];
+static bool events_missed[EVTCHN_2L_NR_CHANNELS];
```

Comment:
Sorry for late review, did not seen this PR before.
If I correctly understood what kind of issues you are facing with this driver, then I think it will be better to add `bool events_missed` member to `struct event_channel_handle` and pass it as a `data` parameter to callback. I think it will simplify this change, especially port calculation :)